<a href="https://colab.research.google.com/github/liuyao12/Ranger-Mish-ImageWoof-5/blob/master/Woof_128_ConvTwist_separable.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ResNet with a Twist

> ConvTwist + Ranger + Mish + MaxBlurPool + restrick

# setup and imports

In [ ]:
# pip install git+https://github.com/kornia/kornia

In [3]:
pip install git+https://github.com/ayasyrev/model_constructor

  Cloning https://github.com/ayasyrev/model_constructor to /tmp/pip-req-build-pg3dhfpa
  Running command git clone -q https://github.com/ayasyrev/model_constructor /tmp/pip-req-build-pg3dhfpa
  Created wheel for model-constructor: filename=model_constructor-0.1.1-cp36-none-any.whl size=23583 sha256=15365fd7e9676dba1db5fff57f6bcdcdcf45d2e649108085a524ba6e476aa89d
  Stored in directory: /tmp/pip-ephem-wheel-cache-_auwmjak/wheels/5b/92/65/8093a307d5802f41f4a8776b40bc12b558e75a2a906ae8b683
Successfully built model-constructor


In [4]:
pip install git+https://github.com/ayasyrev/imagenette_experiments

  Cloning https://github.com/ayasyrev/imagenette_experiments to /tmp/pip-req-build-zaxzzw7s
  Running command git clone -q https://github.com/ayasyrev/imagenette_experiments /tmp/pip-req-build-zaxzzw7s
     |████████████████████████████████| 163kB 4.6MB/s 
  Created wheel for imagenette-experiments: filename=imagenette_experiments-0.0.1-cp36-none-any.whl size=15332 sha256=0101934ec7fdb3fdf6a431d351f2408161be644aae15449874cc66f4448a0e4b
  Stored in directory: /tmp/pip-ephem-wheel-cache-i1ua8gk1/wheels/af/99/98/2212941f45a18bf6d191f56c39e3569138414560c76defc0d4
Successfully built imagenette-experiments
ERROR: kornia 0.3.1 has requirement torch==1.5.0, but you'll have torch 1.5.1+cu101 which is incompatible.


In [5]:
from imagenette_experiments.train_utils import *

In [6]:
from kornia.contrib import MaxBlurPool2d

In [7]:
from fastai.basic_train import *
from fastai.vision import *
# from fastai.script import *
from model_constructor.net import Net, act_fn
from model_constructor.layers import SimpleSelfAttention, ConvLayer

# Twist

In [8]:
class ConvTwist(nn.Module):  # replacing 3x3 Conv2d with in_channels=out_channels
    def __init__(self, ni):
        super(ConvTwist, self).__init__()
        self.twist = True # and n>=128
        self.iters = 1 #3 if ni<=256 else 1
        self.conv = nn.Conv2d(ni, ni, kernel_size=3, padding=1, bias=False, groups=1)
        if self.twist:
            self.expansion = 2
            ni *= self.expansion
            Dx = torch.Tensor([[-1,0,1],[-2,0,2],[-1,0,1]]).view(1,1,3,3)
            Dy = torch.Tensor([[1,2,1],[0,0,0],[-1,-2,-1]]).view(1,1,3,3)
            self.DD = torch.cat([Dx,Dy]*(ni//2), dim=0)
            self.conv1x1 = nn.Conv2d(ni, ni, kernel_size=1, bias=False, groups=1)
            self.mask = torch.zeros((ni,ni,1,1))
            for i in range(ni//2-1):
                self.mask[i*2:i*2+2,i*2:i*2+2] = 1
            self.XY = None

    def forward(self, x):
        N,C,H,W = x.size()
        out = x
        if self.twist:
            CC = C*self.expansion
            double = x.unsqueeze(2).expand(N,C,CC//C,H,W).reshape(N,CC,H,W)
            if self.XY is None:
                XX = torch.from_numpy(np.indices((1,1,H,W))[3]*2/W-1)
                YY = torch.from_numpy(np.indices((1,1,H,W))[2]*2/H-1)
                self.XY = torch.cat([XX,YY]*(CC//2), dim=1).type(x.dtype).to(x.device)
                print(self.XY.size())
                self.DD = self.DD.to(x.device)
                self.mask = self.mask.to(x.device)
            pdf = F.relu(double)+0.01
            total = pdf.sum((2,3))
            XYbar = (self.XY*pdf).sum((2,3))/total
            XY = self.XY-XYbar[:,:,None,None]
            # self.conv1x1.weight.data = self.conv1x1.weight*self.mask # making self.conv1x1 with groups of 2 channels only
            for i in range(self.iters):
                if i>0: double = out.unsqueeze(2).expand(N,C,CC//C,H,W).reshape(N,CC,H,W)
                twist = F.conv2d(double, self.DD, padding=1, groups=CC)
                twist = XY*self.conv1x1(twist)
                out = out + self.conv(out) + twist.view(N,C,CC//C,H,W).sum(2)
            out = out - x
        else:
            out = self.conv(out)
        return out

In [7]:
A = torch.arange(12).view(3,4)
print(A.size())
A.unsqueeze_(1)
print(A.size())
A = A.expand((3,2,4)).reshape(6,4)
print(A)

torch.Size([3, 4])
torch.Size([3, 1, 4])
tensor([[ 0,  1,  2,  3],
        [ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11],
        [ 8,  9, 10, 11]])


# ResBlock

In [9]:
class NewLayer(nn.Sequential):
    """Basic conv layers block"""
    def __init__(self, ni, nf, ks=3, stride=1,
            act=True,  act_fn=nn.ReLU(inplace=True),
            bn_layer=True, bn_1st=True, zero_bn=False,
            padding=None, bias=False, groups=1, **kwargs):

        if padding==None: padding = ks//2
        if ni==nf and ks==3:  layers = [('ConvTwist', ConvTwist(ni))]
        else: layers = [('Conv{}x{}'.format(ks,ks), 
                  nn.Conv2d(ni, nf, ks, stride=stride, padding=padding, bias=bias, groups=groups))]

        act_bn = [('act_fn', act_fn)] if act else []
        if bn_layer:
            bn = nn.BatchNorm2d(nf)
            nn.init.constant_(bn.weight, 0. if zero_bn else 1.)
            act_bn += [('bn', bn)]
        if bn_1st: act_bn.reverse()
        layers += act_bn
        super().__init__(OrderedDict(layers))

In [10]:
class NewResBlock(Module):
    def __init__(self, expansion, ni, nh, stride=1,
                 conv_layer=ConvLayer, act_fn=act_fn, bn_1st=True,
                 pool=nn.AvgPool2d(2, ceil_mode=True), sa=False, sym=False, zero_bn=True):
        nf,ni = nh*expansion,ni*expansion
        conv_layer = NewLayer
        self.reduce = noop if stride==1 else pool
        layers  = [(f"conv_0", conv_layer(ni, nh, 3, act_fn=act_fn, bn_1st=bn_1st)),
                   (f"conv_1", conv_layer(nh, nf, 3, zero_bn=zero_bn, act=False, bn_1st=bn_1st))
        ] if expansion == 1 else [
                   (f"conv_0", conv_layer(ni, nh, 1, act_fn=act_fn, bn_1st=bn_1st)),
                   (f"conv_1", conv_layer(nh, nh, 3, act_fn=act_fn, bn_1st=bn_1st)),
                   (f"conv_2", conv_layer(nh, nf, 1, zero_bn=zero_bn, act=False, bn_1st=bn_1st))
        ]
        if sa: layers.append(('sa', SimpleSelfAttention(nf,ks=1,sym=sym)))
        self.convs = nn.Sequential(OrderedDict(layers))
        self.idconv = noop if ni==nf else conv_layer(ni, nf, 1, act=False, bn_1st=bn_1st)
        self.merge = act_fn

    def forward(self, x):
        o = self.reduce(x)
        return self.merge(self.convs(o) + self.idconv(o))

# Model Constructor

In [11]:
model = Net(c_out=10, layers=[3,6,8,3], expansion=4)
model.block = NewResBlock
# model.conv_layer = NewLayer # for the stem
pool = MaxBlurPool2d(3, True)
model.pool = pool
model.stem_pool = pool
model.stem_sizes = [3,32,64,64]
model.act_fn = Mish()
model.sa = True
res = []

# Runs and results



In [12]:
epochs = [5,5,5]
for e in epochs:
    mixup=0 if e<20 else 0.2
    learn = get_learn(model=model, size=192, bs=32, mixup=mixup)
    learn.fit_fc(e, lr=4e-3, moms=(0.95,0.95), start_pct=0.72)
    res += [learn.recorder.metrics[-1][0].item()]
print([round(x, 6) for x in res], sum(res)/len(res))

data path   /root/.fastai/data/imagewoof2


/usr/local/lib/python3.6/dist-packages/torch/nn/functional.py:2854: UserWarning: The default behavior for interpolate/upsample with float scale_factor will change in 1.6.0 to align with other frameworks/libraries, and use scale_factor directly, instead of relying on the computed output size. If you wish to keep the old behavior, please set recompute_scale_factor=True. See the documentation of nn.Upsample for details. 
  warnings.warn("The default behavior for interpolate/upsample with float scale_factor will change "
/usr/local/lib/python3.6/dist-packages/torch/nn/functional.py:2854: UserWarning: The default behavior for interpolate/upsample with float scale_factor will change in 1.6.0 to align with other frameworks/libraries, and use scale_factor directly, instead of relying on the computed output size. If you wish to keep the old behavior, please set recompute_scale_factor=True. See the documentation of nn.Upsample for details. 
  warnings.warn("The default behavior for interpolate/u

Learn path /root/.fastai/data/imagewoof2


/usr/local/lib/python3.6/dist-packages/torch/nn/functional.py:2854: UserWarning: The default behavior for interpolate/upsample with float scale_factor will change in 1.6.0 to align with other frameworks/libraries, and use scale_factor directly, instead of relying on the computed output size. If you wish to keep the old behavior, please set recompute_scale_factor=True. See the documentation of nn.Upsample for details. 
  warnings.warn("The default behavior for interpolate/upsample with float scale_factor will change "


torch.Size([1, 128, 48, 48])
torch.Size([1, 128, 48, 48])
torch.Size([1, 128, 48, 48])
torch.Size([1, 256, 24, 24])
torch.Size([1, 256, 24, 24])
torch.Size([1, 256, 24, 24])
torch.Size([1, 256, 24, 24])
torch.Size([1, 256, 24, 24])
torch.Size([1, 256, 24, 24])
torch.Size([1, 512, 12, 12])
torch.Size([1, 512, 12, 12])
torch.Size([1, 512, 12, 12])
torch.Size([1, 512, 12, 12])
torch.Size([1, 512, 12, 12])
torch.Size([1, 512, 12, 12])
torch.Size([1, 512, 12, 12])
torch.Size([1, 512, 12, 12])
torch.Size([1, 1024, 6, 6])
torch.Size([1, 1024, 6, 6])
torch.Size([1, 1024, 6, 6])


/pytorch/torch/csrc/utils/python_arg_parser.cpp:756: UserWarning: This overload of addcmul_ is deprecated:
	addcmul_(Number value, Tensor tensor1, Tensor tensor2)
Consider using one of the following signatures instead:
	addcmul_(Tensor tensor1, Tensor tensor2, *, Number value)


data path   /root/.fastai/data/imagewoof2


/usr/local/lib/python3.6/dist-packages/torch/nn/functional.py:2854: UserWarning: The default behavior for interpolate/upsample with float scale_factor will change in 1.6.0 to align with other frameworks/libraries, and use scale_factor directly, instead of relying on the computed output size. If you wish to keep the old behavior, please set recompute_scale_factor=True. See the documentation of nn.Upsample for details. 
  warnings.warn("The default behavior for interpolate/upsample with float scale_factor will change "
/usr/local/lib/python3.6/dist-packages/torch/nn/functional.py:2854: UserWarning: The default behavior for interpolate/upsample with float scale_factor will change in 1.6.0 to align with other frameworks/libraries, and use scale_factor directly, instead of relying on the computed output size. If you wish to keep the old behavior, please set recompute_scale_factor=True. See the documentation of nn.Upsample for details. 
  warnings.warn("The default behavior for interpolate/u

Learn path /root/.fastai/data/imagewoof2


/usr/local/lib/python3.6/dist-packages/torch/nn/functional.py:2854: UserWarning: The default behavior for interpolate/upsample with float scale_factor will change in 1.6.0 to align with other frameworks/libraries, and use scale_factor directly, instead of relying on the computed output size. If you wish to keep the old behavior, please set recompute_scale_factor=True. See the documentation of nn.Upsample for details. 
  warnings.warn("The default behavior for interpolate/upsample with float scale_factor will change "


torch.Size([1, 128, 48, 48])
torch.Size([1, 128, 48, 48])
torch.Size([1, 128, 48, 48])
torch.Size([1, 256, 24, 24])
torch.Size([1, 256, 24, 24])
torch.Size([1, 256, 24, 24])
torch.Size([1, 256, 24, 24])
torch.Size([1, 256, 24, 24])
torch.Size([1, 256, 24, 24])
torch.Size([1, 512, 12, 12])
torch.Size([1, 512, 12, 12])
torch.Size([1, 512, 12, 12])
torch.Size([1, 512, 12, 12])
torch.Size([1, 512, 12, 12])
torch.Size([1, 512, 12, 12])
torch.Size([1, 512, 12, 12])
torch.Size([1, 512, 12, 12])
torch.Size([1, 1024, 6, 6])
torch.Size([1, 1024, 6, 6])
torch.Size([1, 1024, 6, 6])
data path   /root/.fastai/data/imagewoof2


/usr/local/lib/python3.6/dist-packages/torch/nn/functional.py:2854: UserWarning: The default behavior for interpolate/upsample with float scale_factor will change in 1.6.0 to align with other frameworks/libraries, and use scale_factor directly, instead of relying on the computed output size. If you wish to keep the old behavior, please set recompute_scale_factor=True. See the documentation of nn.Upsample for details. 
  warnings.warn("The default behavior for interpolate/upsample with float scale_factor will change "
/usr/local/lib/python3.6/dist-packages/torch/nn/functional.py:2854: UserWarning: The default behavior for interpolate/upsample with float scale_factor will change in 1.6.0 to align with other frameworks/libraries, and use scale_factor directly, instead of relying on the computed output size. If you wish to keep the old behavior, please set recompute_scale_factor=True. See the documentation of nn.Upsample for details. 
  warnings.warn("The default behavior for interpolate/u

Learn path /root/.fastai/data/imagewoof2


torch.Size([1, 128, 48, 48])
torch.Size([1, 128, 48, 48])
torch.Size([1, 128, 48, 48])
torch.Size([1, 256, 24, 24])
torch.Size([1, 256, 24, 24])
torch.Size([1, 256, 24, 24])
torch.Size([1, 256, 24, 24])
torch.Size([1, 256, 24, 24])
torch.Size([1, 256, 24, 24])
torch.Size([1, 512, 12, 12])
torch.Size([1, 512, 12, 12])
torch.Size([1, 512, 12, 12])
torch.Size([1, 512, 12, 12])
torch.Size([1, 512, 12, 12])
torch.Size([1, 512, 12, 12])
torch.Size([1, 512, 12, 12])
torch.Size([1, 512, 12, 12])
torch.Size([1, 1024, 6, 6])
torch.Size([1, 1024, 6, 6])
torch.Size([1, 1024, 6, 6])
[0.796895, 0.789768, 0.787223] 0.7912955085436503


In [13]:
epochs = [20,80]
for e in epochs:
    mixup=0 if e<20 else 0.2
    learn = get_learn(model=model, size=192, bs=32, mixup=mixup)
    learn.fit_fc(e, lr=4e-3, moms=(0.95,0.95), start_pct=0.72)
    res += [learn.recorder.metrics[-1][0].item()]
print([round(x, 6) for x in res], sum(res)/len(res))

data path   /root/.fastai/data/imagewoof2


/usr/local/lib/python3.6/dist-packages/torch/nn/functional.py:2854: UserWarning: The default behavior for interpolate/upsample with float scale_factor will change in 1.6.0 to align with other frameworks/libraries, and use scale_factor directly, instead of relying on the computed output size. If you wish to keep the old behavior, please set recompute_scale_factor=True. See the documentation of nn.Upsample for details. 
  warnings.warn("The default behavior for interpolate/upsample with float scale_factor will change "
/usr/local/lib/python3.6/dist-packages/torch/nn/functional.py:2854: UserWarning: The default behavior for interpolate/upsample with float scale_factor will change in 1.6.0 to align with other frameworks/libraries, and use scale_factor directly, instead of relying on the computed output size. If you wish to keep the old behavior, please set recompute_scale_factor=True. See the documentation of nn.Upsample for details. 
  warnings.warn("The default behavior for interpolate/u

Learn path /root/.fastai/data/imagewoof2


torch.Size([1, 128, 48, 48])
torch.Size([1, 128, 48, 48])
torch.Size([1, 128, 48, 48])
torch.Size([1, 256, 24, 24])
torch.Size([1, 256, 24, 24])
torch.Size([1, 256, 24, 24])
torch.Size([1, 256, 24, 24])
torch.Size([1, 256, 24, 24])
torch.Size([1, 256, 24, 24])
torch.Size([1, 512, 12, 12])
torch.Size([1, 512, 12, 12])
torch.Size([1, 512, 12, 12])
torch.Size([1, 512, 12, 12])
torch.Size([1, 512, 12, 12])
torch.Size([1, 512, 12, 12])
torch.Size([1, 512, 12, 12])
torch.Size([1, 512, 12, 12])
torch.Size([1, 1024, 6, 6])
torch.Size([1, 1024, 6, 6])
torch.Size([1, 1024, 6, 6])
data path   /root/.fastai/data/imagewoof2


/usr/local/lib/python3.6/dist-packages/torch/nn/functional.py:2854: UserWarning: The default behavior for interpolate/upsample with float scale_factor will change in 1.6.0 to align with other frameworks/libraries, and use scale_factor directly, instead of relying on the computed output size. If you wish to keep the old behavior, please set recompute_scale_factor=True. See the documentation of nn.Upsample for details. 
  warnings.warn("The default behavior for interpolate/upsample with float scale_factor will change "
/usr/local/lib/python3.6/dist-packages/torch/nn/functional.py:2854: UserWarning: The default behavior for interpolate/upsample with float scale_factor will change in 1.6.0 to align with other frameworks/libraries, and use scale_factor directly, instead of relying on the computed output size. If you wish to keep the old behavior, please set recompute_scale_factor=True. See the documentation of nn.Upsample for details. 
  warnings.warn("The default behavior for interpolate/u

Learn path /root/.fastai/data/imagewoof2


/usr/local/lib/python3.6/dist-packages/torch/nn/functional.py:2854: UserWarning: The default behavior for interpolate/upsample with float scale_factor will change in 1.6.0 to align with other frameworks/libraries, and use scale_factor directly, instead of relying on the computed output size. If you wish to keep the old behavior, please set recompute_scale_factor=True. See the documentation of nn.Upsample for details. 
  warnings.warn("The default behavior for interpolate/upsample with float scale_factor will change "


torch.Size([1, 128, 48, 48])
torch.Size([1, 128, 48, 48])
torch.Size([1, 128, 48, 48])
torch.Size([1, 256, 24, 24])
torch.Size([1, 256, 24, 24])
torch.Size([1, 256, 24, 24])
torch.Size([1, 256, 24, 24])
torch.Size([1, 256, 24, 24])
torch.Size([1, 256, 24, 24])
torch.Size([1, 512, 12, 12])
torch.Size([1, 512, 12, 12])
torch.Size([1, 512, 12, 12])
torch.Size([1, 512, 12, 12])
torch.Size([1, 512, 12, 12])
torch.Size([1, 512, 12, 12])
torch.Size([1, 512, 12, 12])
torch.Size([1, 512, 12, 12])
torch.Size([1, 1024, 6, 6])
torch.Size([1, 1024, 6, 6])
torch.Size([1, 1024, 6, 6])
[0.796895, 0.789768, 0.787223, 0.878595, 0.893612] 0.8292186379432678


conv: 1-to-1, conv1x1: 2-to-2

in 5, 5, 5, 20, and 80 epochs:
`0.804021, 0.805039, 0.811657, 0.884449, 0.893357`
(2.42 min per epoch)

`0.79944, 0.806058, 0.816238, 0.879359, 0.891066`

`0.796895, 0.789768, 0.787223, 0.878595, 0.893612`


In [14]:
for i in range(80):
    print('epoch {} {}'.format(i, learn.recorder.metrics[i][0].item()))

epoch 0 0.41995418071746826
epoch 1 0.5189615488052368
epoch 2 0.6151692271232605
epoch 3 0.667345404624939
epoch 4 0.6917790770530701
epoch 5 0.736065149307251
epoch 6 0.7696614861488342
epoch 7 0.7630440592765808
epoch 8 0.7900229096412659
epoch 9 0.7946042418479919
epoch 10 0.7956222891807556
epoch 11 0.8086026906967163
epoch 12 0.8170017600059509
epoch 13 0.8162382245063782
epoch 14 0.8340544700622559
epoch 15 0.8376176953315735
epoch 16 0.8429625630378723
epoch 17 0.8421990275382996
epoch 18 0.8462713360786438
epoch 19 0.8470348715782166
epoch 20 0.8449987173080444
epoch 21 0.8488165140151978
epoch 22 0.8574700951576233
epoch 23 0.8546704053878784
epoch 24 0.8368541598320007
epoch 25 0.8640875816345215
epoch 26 0.8523797392845154
epoch 27 0.8539068698883057
epoch 28 0.8495800495147705
epoch 29 0.855942964553833
epoch 30 0.8521252274513245
epoch 31 0.8572155833244324
epoch 32 0.8597607612609863
epoch 33 0.8610333204269409
epoch 34 0.8546704053878784
epoch 35 0.8549249172210693
epoc

In [ ]:
for i in range(80):
    print('epoch {} {}'.format(i, learn.recorder.metrics[i][0].item()))

epoch 0 0.39373886585235596
epoch 1 0.5291422605514526
epoch 2 0.6054975986480713
epoch 3 0.6536014080047607
epoch 4 0.6759989857673645
epoch 5 0.7179943919181824
epoch 6 0.7459913492202759
epoch 7 0.7755153775215149
epoch 8 0.7806057333946228
epoch 9 0.8131840229034424
epoch 10 0.814202070236206
epoch 11 0.8157292008399963
epoch 12 0.8022397756576538
epoch 13 0.8228557109832764
epoch 14 0.8256554007530212
epoch 15 0.8416900038719177
epoch 16 0.8315092921257019
epoch 17 0.8355816006660461
epoch 18 0.8437261581420898
epoch 19 0.8378722071647644
epoch 20 0.8460168242454529
epoch 21 0.8393993377685547
epoch 22 0.8513616919517517
epoch 23 0.8485620021820068
epoch 24 0.8480529189109802
epoch 25 0.8567065596580505
epoch 26 0.8511071801185608
epoch 27 0.8556884527206421
epoch 28 0.8498345613479614
epoch 29 0.853397786617279
epoch 30 0.8612878322601318
epoch 31 0.8569610714912415
epoch 32 0.8605242967605591
epoch 33 0.8615424036979675
epoch 34 0.8518707156181335
epoch 35 0.858488142490387
epoc